sql/src/main/java/org/apache/druid/sql/calcite/expression/DruidExpression.java

```diff
@@ -104,9 +104,24 @@ private static String escape(final String s)
     return escaped.toString();
   }
 
-  public static String numberLiteral(final Number n)
+  public static String longLiteral(final long n)
   {
-    return n == null ? nullLiteral() : n.toString();
+    return String.valueOf(n);
+  }
+
+  public static String doubleLiteral(final double n)
+  {
+    final String s = String.valueOf(n);
+
+    // Ensure number is parsed as a double: add ".0" if necessary.
+    for (int i = 0; i < s.length(); i++) {
+      final char c = s.charAt(i);
+      if (!Character.isDigit(c)) {
```

Comment:
i think you need to check for leading `-` too?

Reply:
This comment is absolutely correct and led me down a crazy rabbit hole 🐰

Reply:
I discovered a couple of things.

First: there is no need to do any of this checking, since simply using `String.valueOf(double)` is enough: it adds the extra `.0` for whole numbers anyway. The real fix to the bug was replacing `numberLiteral` with `longLiteral` and `doubleLiteral`, as this forces us to check the desired type, which means we don't lose the cast. (Previously: we'd lose casts by passing long `3` to `numberLiteral` even though the desired type was DOUBLE.) So, we can simplify the code here, which is nice.

Second: I tried to add a test for `Long.MIN_VALUE` and found that we cannot parse it! It sort of makes sense. To avoid ambiguity between unary minus and negative literals, we don't actually have negative literals. We represent "negative literals" as unary minus next to a positive literal of the absolute value. But! That means `Long.MIN_VALUE` can't be represented, since the absolute value can't fit into a `long` and therefore can't go into a LongExpr. I'm going to try to fix this and see where it leads me. This is where the 🐰 comes in…

Reply:
The latest patch fixes two additional bugs that I noticed in the rabbit hole that this led me down. I updated the PR description to describe all three bugs.